### PR TITLE
fix(react-pi): project streaming updates incrementally

### LINKED ISSUE #7205: react-pi reprojects the full transcript on every streaming frame
## Problem

During streaming, react-pi coalesces updates to animation frames but still rebuilds the projected assistant-ui message array from the full Pi transcript on every frame. It then deep-compares the rebuilt messages to recover shared references.

The cost grows with the entire thread even when only the final streaming message changed.

## Reproduction

A local projection benchmark using the current implementation measured approximately:

| Pi messages | Time per projection |
| ---: | ---: |
| 10 | 0.014 ms |
| 1,000 | 0.66 ms |
| 5,000 | 9.0 ms |

The exact wall time is machine-dependent, but the linear growth is the issue. At 5,000 messages, projection alone consumes more than half of a 16.7 ms frame budget before React renders anything.

## Expected behavior

A streaming update near the end of a long transcript should reuse the unchanged prefix and only reproject the suffix whose grouping can have changed.

## Root cause

`projectPiThreadMessagesShared` calls `projectPiThreadMessages` for the entire input and only performs structural sharing after the full projection has already been created.

## Proposed direction

Pi input messages already retain object identity for the unchanged prefix. We could find the first changed input, move the boundary backward far enough to preserve assistant/tool-result grouping, reuse the corresponding projected prefix, and project only the remaining suffix.

The fix should include a performance contract or benchmark that exercises a long transcript with a changing streaming tail. Please confirm this incremental boundary approach before implementation.

<a href="https://rupic.dev/s/s2.B8lc8itUHrj4PS7Apx3KZX2Ss5DPXNTpcN5h5AoDKvFJs4ffN4u5w3wD24I?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### PR DESCRIPTION
**Pull request by Kinfe123**
## Problem

In `@assistant-ui/react-pi` 0.0.22, every coalesced streaming update projected the full Pi transcript and rebuilt the full exported message repository before structural sharing could reuse anything. Updating only the newest message therefore became slower as the unchanged conversation grew.

Closes #7205.

## Root cause

`projectPiThreadMessagesShared` built a complete tool-result map and converted every source message on each frame. `PiThreadController` then passed every projected message through `ExportedMessageRepository.fromArray` again. The reducer already preserves source-message identity for the unchanged prefix, but neither conversion used that information.

## Change

- retain projection state per thread and locate the first changed source-message identity
- reproject from the owning assistant/tool-result group while preserving tool-result pairing, live tool output, host UI requests, and trailing run status
- maintain tool-result and tool-call indexes so dependent earlier groups are invalidated correctly, including duplicate IDs across a changed suffix
- locate the projected suffix with a binary search, fall back safely if a cached ID breaks the ordering invariant, and reuse converted repository entries before it
- add a seeded 5,000-transition equivalence test covering append, replacement, truncation, tool execution, host UI, and run-status changes
- add a React Pi streaming benchmark with 10, 1,000, and 5,000 stable messages and wire the package into the performance workflow

A focused local run reduced the mean 5,000-message streaming update from about 2.73 ms on the main build to about 0.03 ms on this branch. This benchmark covers the React Pi projection and repository-update layer, not the complete runtime frame. Wall-time numbers are informational; deterministic tests assert unchanged-prefix reuse and equivalence with the full projector. A changed frame still copies top-level array references, but conversion, deep comparison, and repository-entry construction now stay within the dirty suffix.

## Verification

- `pnpm -C packages/react-pi test` (262 tests)
- `pnpm -C packages/react-pi build`
- `pnpm -C packages/x-performance test` (82 tests)
- `pnpm lint`
- `pnpm changesets:check`
- `pnpm size:check`
- focused React Pi benchmark against the built main and branch entries
- React Pi root and node entries remain within the committed size budgets

## Public surface

- `@assistant-ui/react-pi`: behavior-only performance fix with a patch changeset
- no new public exports
- no documentation, API-reference, or template changes
- private performance benchmark and CI wiring added

### CODE REVIEW

File: packages/react-pi/src/runtime/messageProjection.test.ts [hunk 1]
```diff
@@ -420,4 +421,102 @@ describe("messageProjection", () => {
...
+          toolCallId: "tc1",
+          toolName: "bash",
+          content: [{ type: "text", text: "done" }],
+          isError: false,
+          timestamp: 2,
+        },
+      ]),
+    );
+    assertEquivalent(input([user]));
+  });
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.test.ts:481]: Added a seeded 5,000-transition equivalence test. It covers append, replacement, truncation, every currently handled message role plus an unknown role, and mutations to tool executions, host UI requests, and run status. Every transition compares the incremental result with the full projector.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 2]
```diff
@@ -485,6 +499,324 @@ export const projectPiThreadMessagesShared = (
...
+const firstProjectedIndexAtOrAfter = (
+  messages: readonly ThreadMessageLike[],
+  sourceIndex: number,
+) => {
+  const index = messages.findIndex((message) => {
+    const projectedIndex = projectedSourceIndex(message);
+    return projectedIndex !== undefined && projectedIndex >= sourceIndex;
+  });
+  return index === -1 ? messages.length : index;
+};
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts]: Addressed. The projected-boundary lookup now uses binary search, and the controller reuses converted repository entries before the changed suffix. The PR description also calls out the remaining top-level array reference copies. On the final built branch, the focused 5,000-message mean is about 0.03 ms versus about 2.73 ms on the main build.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 3]
```diff
@@ -485,6 +499,324 @@ export const projectPiThreadMessagesShared = (
...
+
+export class PiThreadMessageProjector {
+  private previousInput: PiProjectionInput | undefined;
+  private projectedMessages: readonly ThreadMessageLike[] = [];
+  private toolResults = new Map<string, ProjectedToolResult>();
+  private toolCallIndices = new Map<string, number>();
+
+  public constructor(projectedMessages: readonly ThreadMessageLike[] = []) {
+    this.projectedMessages = projectedMessages;
+  }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts:727]: Addressed. The constructor parameter is removed, and the controller now initializes the projector and its initial projection explicitly in the constructor, so there is no class-field ordering dependency.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 4]
```diff
@@ -485,6 +499,324 @@ export const projectPiThreadMessagesShared = (
...
+    }
+
+    let dirtyIndex = messageChangeIndex;
+    if (messageChangeIndex !== undefined) {
+      for (const index of [
+        trailingAssistantMessageIndex(previousInput.messages),
+        trailingAssistantMessageIndex(input.messages),
+      ]) {
+        if (index !== undefined) {
+          dirtyIndex = Math.min(dirtyIndex, index);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts:770]: This is already narrowed safely. dirtyIndex is initialized from messageChangeIndex, and this call is inside the messageChangeIndex !== undefined branch, so TypeScript treats it as a number. The strict package build and CI Build Changed Packages both compile this line successfully. No code change is needed.

File: packages/x-performance/bench/react-pi-message-projection.bench.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        contentIndex: 0,
+        delta: "x",
+        partial: message,
+      },
+    });
+    scheduled.shift()?.();
+  };
+};
+
+describe("react-pi: streaming tail projection", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/x-performance/bench/react-pi-message-projection.bench.ts:88]: Vitest supports an async suite factory here. I verified this with the committed file: vitest bench registered and executed all three rows for 10, 1,000, and 5,000 stable messages. The performance workflow uses the same Vitest benchmark runner, so moving setup to beforeAll is not needed.

File: packages/react-pi/src/runtime/messageProjection.test.ts [hunk 6]
```diff
@@ -420,4 +421,246 @@ describe("messageProjection", () => {
     expect(contentParts(out[0]!)[0]!.approval).toBeUndefined();
     expect(out[0]!.status).toEqual({ type: "complete", reason: "stop" });
   });
+
+  it("keeps incremental projection equivalent across external state changes", () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.test.ts:425]: Addressed in 22f42e4. The committed deterministic probe runs 5,000 transitions across appends, replacements, truncations, the supported message roles, an unknown role, tool executions, host UI requests, and run status, comparing the incremental and full projectors after every transition.

File: packages/react-pi/src/runtime/messageProjection.test.ts [hunk 7]
```diff
@@ -420,4 +421,233 @@ describe("messageProjection", () => {
...
+          };
+        case 5:
+          return {
+            role: "custom",
+            customType: "test",
+            content: `custom-${index}`,
+            display: index % 2 === 0,
+            timestamp: index,
+          };
+        default:
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.test.ts]: Addressed in 52af2bc. The seeded generator now uses nine outcomes, including an explicit compactionSummary branch and a future-role value that exercises the projector default path. The focused package suite passes with both cases included.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 8]
```diff
@@ -479,11 +493,337 @@ export const shareProjectedThreadMessages = (
...
+  next: readonly PiAgentMessage[],
+  startIndex: number,
+) => {
+  for (let index = startIndex; index < previous.length; index++) {
+    const message = previous[index]!;
+    if (message.role !== "assistant") continue;
+    for (const part of (message as PiAssistantMessage).content) {
+      if (part.type === "toolCall") indices.delete(part.id);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts:577]: Fixed in 4caf03e2d. Tool-call index updates now restore the latest matching call from the retained prefix when a duplicate disappears from the suffix. I also added a regression test that removes a later duplicate and then updates the earlier call through live tool execution state.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 9]
```diff
@@ -479,11 +493,337 @@ export const shareProjectedThreadMessages = (
...
+const firstProjectedIndexAtOrAfter = (
+  messages: readonly ThreadMessageLike[],
+  sourceIndex: number,
+) => {
+  let low = 0;
+  let high = messages.length;
+  while (low < high) {
+    const middle = (low + high) >>> 1;
+    const projectedIndex = projectedSourceIndex(messages[middle]!);
+    if (projectedIndex === undefined || projectedIndex >= sourceIndex) {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts]: Fixed in 4caf03e2d. An unparseable cached projected ID now returns a fallback signal, and the projector rebuilds from source index 0 instead of retaining or dropping a partial prefix. The new regression test injects an unexpected cached ID and confirms the result still matches full projection.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 10]
```diff
@@ -479,11 +493,370 @@ export const shareProjectedThreadMessages = (
...
+const firstProjectedIndexAtOrAfter = (
+  messages: readonly ThreadMessageLike[],
+  sourceIndex: number,
+): number | undefined => {
+  let low = 0;
+  let high = messages.length;
+  while (low < high) {
+    const middle = (low + high) >>> 1;
+    const projectedIndex = projectedSourceIndex(messages[middle]!);
+    if (projectedIndex === undefined) return undefined;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts]: Fixed in 54dff64d4. The projector now validates the complete projected ID sequence when it builds that sequence and caches the parsed source indices privately. Later updates binary-search only that validated cache and update its dirty suffix. If validation cannot establish strict ordering, the cache is cleared and the next update rebuilds from source index 0. The regression test covers the fallback, and the focused 5,000-message benchmark remains about 0.034 ms locally.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 11]
```diff
@@ -479,11 +493,409 @@ export const shareProjectedThreadMessages = (
...
+      const suffixSourceIndices = validatedProjectedSourceIndices(
+        nextSuffix,
+        this.projectedSourceIndices[projectedStartIndex - 1],
+      );
+      if (suffixSourceIndices) {
+        this.projectedSourceIndices.splice(
+          projectedStartIndex,
+          this.projectedSourceIndices.length - projectedStartIndex,
+          ...suffixSourceIndices,
+        );
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts:875]: Fixed in ba7959203. The cache now truncates the old suffix with splice and appends each validated index without spreading a wide array into a function call. The focused tests, package build, size check, and 5,000-message benchmark all pass; the benchmark remains about 0.038 ms locally.

File: packages/react-pi/src/runtime/messageProjection.ts [hunk 12]
```diff
@@ -479,11 +493,411 @@ export const shareProjectedThreadMessages = (
...
+        const index = this.toolCallIndices.get(id);
+        if (index !== undefined) {
+          dirtyIndex =
+            dirtyIndex === undefined ? index : Math.min(dirtyIndex, index);
+        }
+      }
+    }
+
+    if (previousInput.runStatus !== input.runStatus) {
+      const index = lastAssistantMessageIndex(input.messages);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-pi/src/runtime/messageProjection.ts]: Fixed in c84760aae. Run-status changes now use trailingAssistantMessageIndex, so transcripts ending in a user, bash execution, or another standalone role do not enter the projection path. The obsolete broader helper was removed, and focused tests, build, lint, formatting, and size checks pass.
